pallets/parachain-staking/src/lib.rs

```diff
@@ -0,0 +1,1295 @@
+		fn execute_delayed_collator_exits(next: RoundIndex) {
+			let remain_exits = <ExitQueue<T>>::get()
+				.into_iter()
+				.filter_map(|x| {
+					if x.amount > next {
+						Some(x)
+					} else {
+						if let Some(state) = <CollatorState<T>>::get(&x.owner) {
+							for bond in state.delegators.into_iter() {
+								// return stake to delegator
+								// NOTE: Since this function will be removed anyway, the below is okay for now
+								Self::prepare_withdraw(&bond.owner, bond.amount).ok()?;
+								// remove delegation from delegator state
+								if let Some(mut delegator) = <DelegatorState<T>>::get(&bond.owner) {
+									if let Some(remaining) = delegator.rm_delegation(x.owner.clone()) {
+										if remaining.is_zero() {
+											<DelegatorState<T>>::remove(&bond.owner);
+										} else {
+											<DelegatorState<T>>::insert(&bond.owner, delegator);
+										}
+									}
+								}
+							}
+							// return stake to collator
+							// NOTE: Since this function will be removed anyway, the below is okay for now
+							Self::prepare_withdraw(&state.id, state.bond).ok()?;
+
+							let TotalStake {
+								collators: total_collators,
+								delegators: total_delegators,
+							} = <Total<T>>::get();
+							let total_collators = total_collators.saturating_sub(state.bond);
+							// safe because bond <= total at all times
+							let total_delegators = total_delegators.saturating_sub(state.total - state.bond);
+							<Total<T>>::put(TotalStake {
+								collators: total_collators,
+								delegators: total_delegators,
+							});
+
+							<CollatorState<T>>::remove(&x.owner);
+							Self::deposit_event(Event::CollatorLeft(
+								x.owner,
+								state.total,
+								total_collators,
+								total_delegators,
+							));
+						}
+						None
+					}
+				})
+				.collect::<Vec<Bond<T::AccountId, RoundIndex>>>();
+			<ExitQueue<T>>::put(OrderedSet::from(remain_exits));
+		}
+
+		/// Select the top `n` collators in terms of cumulated stake (self +
+		/// from delegators) from the CandidatePool to become block authors for
+		/// the next round.
+		fn select_top_candidates(next: RoundIndex) -> (u32, BalanceOf<T>, BalanceOf<T>) {
+			let (mut all_collators, mut total_collators, mut total_delegators) =
+				(0u32, BalanceOf::<T>::zero(), BalanceOf::<T>::zero());
+			log::trace!("Select collators for round {}", next);
+			let mut candidates = <CandidatePool<T>>::get().to_vec();
+			let top_n = <TotalSelected<T>>::get() as usize;
+
+			log::trace!("{} Candidates for {} Collator seats", candidates.len(), top_n);
+
+			// Order candidates by their total stake
+			candidates.sort_unstable_by(|a, b| a.amount.cmp(&b.amount));
+
+			// Choose the top TotalSelected qualified candidates, ordered by stake (least to
+			// greatest, thus requires `rev()`)
+			let mut collators = candidates
+				.into_iter()
+				.rev()
+				.take(top_n)
+				.filter(|x| x.amount >= T::MinCollatorStk::get())
```

Comment:
What happens if any of the `top_n` has less than the minimum amount? I guess this is also impossible if the other functions are correctly implemented, but would be good to have this highlighted here.